File: src/functionalTest/resources/dsl-coverage/mysql-liquibase--override-all-props/src/main/java/com/example/UserController.java
package com.example;

import com.example.jooq.Tables;
import com.example.jooq.tables.records.AddressesRecord;
import com.example.jooq.tables.records.UsersRecord;
import java.math.BigInteger;
import java.sql.Date;
import java.util.List;
import org.jooq.DSLContext;

public class UserController {

    public record AddressDto(Integer id, String country, String zip, String city, String street) {}
    public record UserDto(Integer id, String name, BigInteger pointsAmount, Date dateOfBirth, List<AddressDto> addresses) {}

    private final DSLContext dsl;

    public UserController(DSLContext dsl) {
        this.dsl = dsl;
    }

    public UserDto createUser(UserDto request) {
        UsersRecord user = dsl.insertInto(Tables.USERS)
            .set(Tables.USERS.NAME, request.name())
            .set(Tables.USERS.POINTS_AMOUNT, request.pointsAmount())
            .set(Tables.USERS.DATE_OF_BIRTH, request.dateOfBirth())
            .returning()
            .fetchOne();

        List<AddressesRecord> addresses = request.addresses().stream()
            .map(a -> dsl.insertInto(Tables.ADDRESSES)
                .set(Tables.ADDRESSES.USER_ID, user.getId())
                .set(Tables.ADDRESSES.COUNTRY, a.country())
                .set(Tables.ADDRESSES.ZIP, a.zip())
                .set(Tables.ADDRESSES.CITY, a.city())
                .set(Tables.ADDRESSES.STREET, a.street())
                .returning()
                .fetchOne())
            .toList();

        return toDto(user, addresses);
    }

    public UserDto getUser(int id) {
        UsersRecord user = dsl.fetchOne(Tables.USERS, Tables.USERS.ID.eq(id));
        List<AddressesRecord> addresses = dsl.fetch(Tables.ADDRESSES, Tables.ADDRESSES.USER_ID.eq(id));
        return toDto(user, addresses);
    }

    private UserDto toDto(UsersRecord user, List<AddressesRecord> addresses) {
        return new UserDto(
            user.getId(),
            user.getName(),
            user.getPointsAmount(),
            user.getDateOfBirth(),
            addresses.stream()
                .map(a -> new AddressDto(a.getId(), a.getCountry(), a.getZip(), a.getCity(), a.getStreet()))
                .toList()
        );
    }
}

File: src/functionalTest/resources/spring-4.0/h2-flyway--kotlin/src/main/kotlin/com/example/UserController.kt
package com.example

import com.example.jooq.tables.records.AddressesRecord
import com.example.jooq.tables.records.UsersRecord
import com.example.jooq.tables.references.ADDRESSES
import com.example.jooq.tables.references.USERS
import org.jooq.DSLContext
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/users")
class UserController(private val dsl: DSLContext) {

    data class AddressDto(val id: Int?, val country: String, val zip: String, val city: String, val street: String)
    data class UserDto(val id: Int?, val name: String, val pointsAmount: Long?, val dateOfBirth: java.time.LocalDate?, val addresses: List<AddressDto>)

    @PostMapping
    fun createUser(@RequestBody request: UserDto): UserDto {
        val user = dsl.insertInto(USERS)
            .set(USERS.NAME, request.name)
            .set(USERS.POINTS_AMOUNT, request.pointsAmount)
            .set(USERS.DATE_OF_BIRTH, request.dateOfBirth)
            .returning()
            .fetchOne()!!

        val addresses = request.addresses.map { a ->
            dsl.insertInto(ADDRESSES)
                .set(ADDRESSES.USER_ID, user.id)
                .set(ADDRESSES.COUNTRY, a.country)
                .set(ADDRESSES.ZIP, a.zip)
                .set(ADDRESSES.CITY, a.city)
                .set(ADDRESSES.STREET, a.street)
                .returning()
                .fetchOne()!!
        }

        return toDto(user, addresses)
    }

    @GetMapping("/{id}")
    fun getUser(@PathVariable id: Int): UserDto {
        val user = dsl.fetchOne(USERS, USERS.ID.eq(id))!!
        val addresses = dsl.fetch(ADDRESSES, ADDRESSES.USER_ID.eq(id))
        return toDto(user, addresses)
    }

    private fun toDto(user: UsersRecord, addresses: List<AddressesRecord>): UserDto =
        UserDto(
            id = user.id,
            name = user.name!!,
            pointsAmount = user.pointsAmount,
            dateOfBirth = user.dateOfBirth,
            addresses = addresses.map { a ->
                AddressDto(a.id, a.country!!, a.zip!!, a.city!!, a.street!!)
            }
        )
}
